### PR TITLE
(manifests/v2) add support for webhooks by default and docs

### LINKED ISSUE #4240: Document: Run pkg manifest does not work with webhooks and certmanager
#### What did you do?

Try to run the Memcached Go testdata project which has a webhook implementation with `run packagemanifests`  

#### What did you expect to see?

All worked well as expected. 

#### What did you see instead? Under which circumstances?

Failed to mount the cert-manager: 

```
6s          Warning   FailedMount               pod/memcached-operator-controller-manager-55f78f598d-wprvr    MountVolume.SetUp failed for volume "cert" : secret "webhook-server-cert" not found
```

Then, if we remove the https://github.com/operator-framework/operator-sdk/blob/master/testdata/go/memcached-operator/bundle/manifests/memcached-operator.clusterserviceversion.yaml#L140-L149 as a workaround the manager is nerver deployed/installed:

```
    Unexpected error:
        <*errors.errorString | 0xc000298480>: {
            s: "operator-sdk run packagemanifests --install-mode AllNamespaces --version 0.0.1 --timeout 4m failed with error: time=\"2020-11-16T16:16:33Z\" level=info msg=\"Creating memcached-operator registry\"\ntime=\"2020-11-16T16:16:33Z\" level=info msg=\"  Creating ConfigMap \\\"default/memcached-operator-registry-manifests-package\\\"\"\ntime=\"2020-11-16T16:16:33Z\" level=info msg=\"  Creating ConfigMap \\\"default/memcached-operator-registry-manifests-0-0-1\\\"\"\ntime=\"2020-11-16T16:16:33Z\" level=info msg=\"  Creating Deployment \\\"default/memcached-operator-registry-server\\\"\"\ntime=\"2020-11-16T16:16:33Z\" level=info msg=\"  Creating Service \\\"default/memcached-operator-registry-server\\\"\"\ntime=\"2020-11-16T16:16:33Z\" level=info msg=\"Waiting for Deployment \\\"default/memcached-operator-registry-server\\\" rollout to complete\"\ntime=\"2020-11-16T16:16:33Z\" level=info msg=\"  Waiting for Deployment \\\"default/memcached-operator-registry-server\\\" to rollout: 0 of 1 updated replicas are available\"\ntime=\"2020-11-16T16:16:39Z\" level=info msg=\"  Deployment \\\"default/memcached-operator-registry-server\\\" successfully rolled out\"\ntime=\"2020-11-16T16:16:39Z\" level=info msg=\"Created CatalogSource: memcached-operator-catalog\"\ntime=\"2020-11-16T16:16:39Z\" level=info msg=\"OperatorGroup \\\"operator-sdk-og\\\" created\"\ntime=\"2020-11-16T16:16:39Z\" level=info msg=\"Created Subscription: memcached-operator-v0-0-1-sub\"\ntime=\"2020-11-16T16:16:43Z\" level=info msg=\"Approved InstallPlan install-8k225 for the Subscription: memcached-operator-v0-0-1-sub\"\ntime=\"2020-11-16T16:16:43Z\" level=info msg=\"Waiting for ClusterServiceVersion \\\"default/memcached-operator.v0.0.1\\\" to reach 'Succeeded' phase\"\ntime=\"2020-11-16T16:16:43Z\" level=info msg=\"  Waiting for ClusterServiceVersion \\\"default/memcached-operator.v0.0.1\\\" to appear\"\ntime=\"2020-11-16T16:16:47Z\" level=info msg=\"  Found ClusterServiceVersion \\\"default/memcached-operator.v0.0.1\\\" phase: Pending\"\ntime=\"2020-11-16T16:16:48Z\" level=info msg=\"  Found ClusterServiceVersion \\\"default/memcached-operator.v0.0.1\\\" phase: InstallReady\"\ntime=\"2020-11-16T16:16:49Z\" level=info msg=\"  Found ClusterServiceVersion \\\"default/memcached-operator.v0.0.1\\\" phase: Installing\"\ntime=\"2020-11-16T16:20:33Z\" level=fatal msg=\"Failed to run packagemanifests: error waiting for CSV to install: timed out waiting for the condition\\n\"\n",
        }
        operator-sdk run packagemanifests --install-mode AllNamespaces --version 0.0.1 --timeout 4m failed with error: time="2020-11-16T16:16:33Z" level=info msg="Creating memcached-operator registry"
        time="2020-11-16T16:16:33Z" level=info msg="  Creating ConfigMap \"default/memcached-operator-registry-manifests-package\""
        time="2020-11-16T16:16:33Z" level=info msg="  Creating ConfigMap \"default/memcached-operator-registry-manifests-0-0-1\""
        time="2020-11-16T16:16:33Z" level=info msg="  Creating Deployment \"default/memcached-operator-registry-server\""
        time="2020-11-16T16:16:33Z" level=info msg="  Creating Service \"default/memcached-operator-registry-server\""
        time="2020-11-16T16:16:33Z" level=info msg="Waiting for Deployment \"default/memcached-operator-registry-server\" rollout to complete"
        time="2020-11-16T16:16:33Z" level=info msg="  Waiting for Deployment \"default/memcached-operator-registry-server\" to rollout: 0 of 1 updated replicas are available"
        time="2020-11-16T16:16:39Z" level=info msg="  Deployment \"default/memcached-operator-registry-server\" successfully rolled out"
        time="2020-11-16T16:16:39Z" level=info msg="Created CatalogSource: memcached-operator-catalog"
        time="2020-11-16T16:16:39Z" level=info msg="OperatorGroup \"operator-sdk-og\" created"
        time="2020-11-16T16:16:39Z" level=info msg="Created Subscription: memcached-operator-v0-0-1-sub"
        time="2020-11-16T16:16:43Z" level=info msg="Approved InstallPlan install-8k225 for the Subscription: memcached-operator-v0-0-1-sub"
        time="2020-11-16T16:16:43Z" level=info msg="Waiting for ClusterServiceVersion \"default/memcached-operator.v0.0.1\" to reach 'Succeeded' phase"
        time="2020-11-16T16:16:43Z" level=info msg="  Waiting for ClusterServiceVersion \"default/memcached-operator.v0.0.1\" to appear"
        time="2020-11-16T16:16:47Z" level=info msg="  Found ClusterServiceVersion \"default/memcached-operator.v0.0.1\" phase: Pending"
        time="2020-11-16T16:16:48Z" level=info msg="  Found ClusterServiceVersion \"default/memcached-operator.v0.0.1\" phase: InstallReady"
        time="2020-11-16T16:16:49Z" level=info msg="  Found ClusterServiceVersion \"default/memcached-operator.v0.0.1\" phase: Installing"
        time="2020-11-16T16:20:33Z" level=fatal msg="Failed to run packagemanifests: error waiting for CSV to install: timed out waiting for the condition\n"

```  

#### Environment

**Operator type:**
/language go

**Kubernetes cluster type:**

$ operator-sdk version; eb694f2
$ go version: 1.15

#### Possible Solution

I tried the following workaround but it does not work as well. 

- We would have to specify the webhook server details while setting it up in controller runtime (ie) https://github.com/varshaprasad96/memcached-operator/blob/5cf47801493e7e9075bacce9f10595350a33b09d/api/v1/memcached_webhook.go#L38-L49
- Also there are a few minor changes required in CSV: https://github.com/varshaprasad96/memcached-operator/blob/5cf47801493e7e9075bacce9f10595350a33b09d/packagemanifests/0.0.1/memcached-operator.clusterserviceversion.yaml#L219-L261
- Most of it is already fixed from SDK’s end in Master. But the only thing is while using olm for conversion webhooks, we need to comment the volumeMounts and volumes since OLM will place the certificate in another location and we dont want it to look at the cert-managers mountPath which we generally have.

- This is a working memcached example: https://github.com/varshaprasad96/memcached-operator/tree/conv-wh-specify-config

#### Additional context

- OLM currently has the fix to place the cert in the location specified by controller-runtime with this PR: https://github.com/operator-framework/operator-lifecycle-manager/pull/1808
- And this is the issue which tracks support for cert-manager and self-signed certs to be specified in CSV: https://github.com/operator-framework/operator-lifecycle-manager/issues/1805

### LINKED ISSUE #4244: How to get webhook + OLM generation right
## Type of question

How to implement a specific feature

#### What did you do?

I was using `v1.0.0` to generate a webhook that I was using with OLM. I had to add in several "hacks"
to get around some issues, namely the webhook container port, target port, and deployment name. To do this
after building the bundle I would manipulate the CSV using `yq` to set these values appropriately.

```
    yq w -i "${csv}" spec.webhookdefinitions[*].containerPort "9443" 
    yq w -i "${csv}" spec.webhookdefinitions[*].targetPort "9443" 
    yq w -i "${csv}" spec.webhookdefinitions[*].deploymentName "<kustomize-prefix>-controller-manager" 
```

This lead to the created ValidatingWebhookConfiguration pointing at `<kustomize-prefix>-controller-manager-service` on port 9443 and a service being created called `<kustomize-prefix>-controller-manager-service` with port 9443, target port 9443. And everything lined up and worked as expected

I have just moved up to version `v1.2.0` which says it fixes the issues for which I had hacks above. So I removed the hacks I had in place and reran.

#### What am I seeing?

I am seeing behaviour I am confused by, and am now not sure how I am supposed to set this up so it works correctly.

I am really struggling to put into words what is being created and how it all isn't quite fitting together, because I am not sure what the expected behaviour is now. So I think it would be really helpful if someone could explain what it is that is expected to be generated and with what values, and what files those values are based on. So that I can compare and then say what I have got that is different from what is expected.

The essence of the problem is, since moving to `1.2.0` when I install the operator through OLM and remove the hacks I had in place. where I had 1 9443:9443 service (`<kustomize-prefix>-controller-manager-service`) and the generated `ValidatingWebhookConfiguration` was pointing at that service,
I know have 2 services. A `443:443` one (`<kustomize-prefix>-controller-manager-service`) that the webhook is pointing at, which therefore does not work, and a `443:9443` (`<kustomize-prefix>-webhook-service`) one that the webhook is not pointing at so is redundant.

I have tried to manipulate values to bring things in line, but everything I change just seems to make things worse or has no impact at all

**Operator type:**

/language go

**Kubernetes cluster type:**  OpenShift

`$ operator-sdk version` 1.2.0

`$ go version ` 1.15

### LINKED ISSUE #4439: Webhook mount location is incompatible with older versions of OLM
#### What did you do?

Attempt to deploy an operator with webhooks generated by SDK v1.1.0 against the `release-4.6` version of OLM.

#### What did you expect to see?

The operator falls back to the _old_ default serving cert mount locations for the `release-4.6` version of OLM if the _newer_  locations aren't populated: 

_older_ `release-4.6` locations:

- `/apiserver.local.config/certificates/apiserver.crt`
- `/apiserver.local.config/certificates/apiserver.key`

_newer_ `master` locations:

- `/tmp/k8s-webhook-server/serving-certs/tls.cert`
- `/tmp/k8s-webhook-server/serving-certs/tls.key`

#### What did you see instead? Under which circumstances?

The operator attempted to use _only_ the new, unpopulated, serving cert mount locations and failed to be deployed:

```sh
$ kubectl get events -n operators

55s         Warning   Failed                pod/opentelemetry-operator-controller-manager-9bdd9fcc4-ctr8k     Error: cannot find volume "apiservice-cert" to mount into container "kube-rbac-proxy"
55s         Warning   Failed                pod/opentelemetry-operator-controller-manager-9bdd9fcc4-ctr8k     Error: cannot find volume "apiservice-cert" to mount into container "manager"
```

#### Environment

**Operator type:**

/language go

**Kubernetes cluster type:**

OpenShift

`$ operator-sdk version`

v1.1.0

#### Possible Solution

- Fall back to older default mount locations when current locations aren't populated
- Document OLM compatibility

#### Additional context

- A patch has been introduced into OLM that mounts serving certs in both the old and new locations. This **will not** be backported to 4.6 or any earlier releases since it would amount to supporting _forwards compatibility_ with SDK.
- [Original Issue](https://github.com/operator-framework/operator-lifecycle-manager/issues/1839)
- [Issue: documenting SDK/OLM compatibility](https://github.com/operator-framework/operator-sdk/issues/4438)

### PR DESCRIPTION
**Pull request by estroz**
**Description of the change:**
- internal/plugins/manifests/v2: initialize project with a `config/manifests/kustomization.yaml` that contains a patch to remove the [`cert` volume and mount](https://github.com/operator-framework/operator-sdk/blob/a07364f06aa14934ce52257f3af1aaf0da587401/testdata/go/v3/memcached-operator/config/default/manager_webhook_patch.yaml) from the default manager configuration.
- internal/cmd/operator-sdk/generate/kustomize: use the manifests/v2 kustomization.yaml for backwards-compatibility.

**Motivation for the change:** no kustomize helpers are created to enable webhooks by default. This is considered a bug since the CSV generator supports webhooks (see #4244).

Closes #4240 
Closes #4244
Closes #4439 

/kind bug

**Checklist**

If the pull request includes user-facing changes, extra documentation is required:
- [x] Add a new changelog fragment in `changelog/fragments` (see [`changelog/fragments/00-template.yaml`](https://github.com/operator-framework/operator-sdk/tree/master/changelog/fragments/00-template.yaml))
- [x] Add or update relevant sections of the docs website in [`website/content/en/docs`](https://github.com/operator-framework/operator-sdk/tree/master/website/content/en/docs)

### CODE REVIEW

**Comment by estroz** [comment]
Still need to update docs and add a changelog

File: testdata/go/v3/memcached-operator/config/manifests/kustomization.yaml
Comment (camilamacedo86) [inline testdata/go/v3/memcached-operator/config/manifests/kustomization.yaml:10]: Question: Have this need already tracked on the OLM side?

**Review by camilamacedo86 — COMMENTED** [review]
IMO we are just missing some nits to get this merged:
- We need to provide the changelog which also will be useful for those who face this problem. 
- Would be nice to add the manual steps to fix that in the FAQ as well. Do we have them already?

File: testdata/go/v3/memcached-operator/config/manifests/kustomization.yaml
Comment (varshaprasad96) [reply testdata/go/v3/memcached-operator/config/manifests/kustomization.yaml:10]: I think https://github.com/operator-framework/operator-lifecycle-manager/issues/1805 was created for this

File: hack/generate/samples/internal/go/v2/memcached_with_webhooks.go
Comment (camilamacedo86) [inline hack/generate/samples/internal/go/v2/memcached_with_webhooks.go:94]: Our goal is to integrate the projects with OLM. So, I think is very fine we try to solve that by default. 
However, I'd like to raise some concerns about how we are doing that. 

- Why we do not change the sdk command which generates the OLM file to remove the amount from the bundle instead of changing the default scaffold? Is not it only a problem for OLM since it is not supported? 
- Why are we adding things that need to be removed manually after when the CERT starts to be supported? Why we are increasing the complexity of the kustomize file instead of only solve the issue for / where / when it is a problem until we are able to address this need? then, see that when we start to support cert-manager we will not need for the users to change their projects manually again. 
- WDYT about when the bundle be generated and the amount be removed it be informed to the user by a warning output with the link for the issue? IMO we need to clarify that it is not supported and provide a way for the follow up the issue as well. 

/hold

WDYT @estroz @varshaprasad96 @jmrodri

Comment (estroz) [reply hack/generate/samples/internal/go/v2/memcached_with_webhooks.go:94]: * The mount is only a problem for OLM, since it does not support cert-manager yet. Changing `config/default` would alter kubebuilder behavior, which we don't want. The whole point of kustomize is to create a hierarchy of YAML files and combine them in different ways for different environments, which `config/manifests` does with minimal changes.
* This is a problem _now_. Webhooks do not work when deployed via OLM. See https://github.com/operator-framework/operator-sdk/issues/4244#issuecomment-732355073.
* The UX you suggest is way worse than just patching `kustomize build config/default` output. Plus that would break "vanilla" operator deployment. OLM already [documents](https://olm.operatorframework.io/docs/advanced-tasks/adding-admission-and-conversion-webhooks/#certificate-authority-requirements) that it doesn't support cert-manager. This will be linked in the docs I intend to write before this PR is merged.

>I think we need to be very careful by addressing workarounds that require manual changes about changes that affect the files and users generated by the tool and the users. Have we not another way to solve that?

There is no other way to solve this problem short of OLM supporting cert-manager _with minimal changes_. You could replicate `config/default` in `config/manifests` (indeed, I [tried that](https://github.com/estroz/operator-sdk/commit/50c90f85ecd727a738b3c64ffeb535c99a5ff721)), but this path creates more places for users to update YAML despite being more kustomize-y.

Comment (camilamacedo86) [reply hack/generate/samples/internal/go/v2/memcached_with_webhooks.go:94]: Can we not make the opertor-sdk generate bundle remove it when gen the csv?

Comment (estroz) [reply hack/generate/samples/internal/go/v2/memcached_with_webhooks.go:94]: Then `generate bundle` is (poorly) emulating kustomize behavior, more so than the command already is. How will it know which mount to remove if a user changes its name? What if more cert-manager “stuff” is added later to kustomize configs that also needs to be removed?

Comment (camilamacedo86) [reply hack/generate/samples/internal/go/v2/memcached_with_webhooks.go:94]: Fair point 👍 
/hold cancel

**Review by camilamacedo86 — APPROVED** [review]
/lgtm 
/approved

**Comment by camilamacedo86** [comment]
/hold

It is true. It is missing the changelog and docs.

**Review by jmrodri — APPROVED** [review]
/lgtm

**Pull request merged**
